### DISCUSSION TITLE
conflicted interdiff has counter-intuitive output

### DISCUSSION THREAD

**Discussion by Zoybean** [jj-vcs/jj#4246] [Ideas]
In an interdiff between 2 commits that do not resolve any conflicts, and that do not agree on the initial state of a modified file, the output states "conflict resolved in \<file\>". This is not necessarily problematic, but it was surprising, and I think it could probably be improved.

Here's a fairly minimal example of what I mean:
1. `jj git init`
2. `echo "foo" > foo`
3. `jj branch create foo`
4. `jj new`
5. `echo "bar" > foo`
7. `jj new`
8. `echo "baz" > foo`
9. `jj branch create baz`
10. `jj new foo`
11. `echo "qux" > foo`
12. `jj branch create qux`
13. `jj interdiff --from baz --to qux`

output is:
```
Resolved conflict in foo:
   1     : <<<<<<< Conflict 1 of 1
   2     : %%%%%%% Changes from base to side #1
   3     : -bar
   4     : +foo
   5     : +++++++ Contents of side #2
   6     : baz
   7    1: >>>>>>> Conflict 1 of 1 endsqux
```

I don't really know what the output should say, which is why I haven't made this a proper bug report or feature request. It just seems that the current output is misleading, as there is no actual resolved conflict in an interdiff. I thought it *was* a bug, initially, until I managed to find a minimal example and realised it was just an ordinary conflict in an interdiff.

**Comment by ilyagr** [comment]
Interesting point. Thanks for raising it!

I'm a bit flummoxed -- I see how this can be confusing, but I'm also having a hard time seeing how this could be redesigned to work better. I find `interdiff` a bit confusing in general, and your example is an excellent illustration for that, but I don't have a better idea for its use-case.

--------

You may have already figured this out, but to clarify what's going on: `jj interdiff --from baz --to qux` is designed give similar output to if you rebased and then did the diff. So, it's similar to `jj rebase -r baz --to qux-; jj diff --from baz --to qux`, except that `interdiff` does this without actually moving the commit.

In your example, the rebase would have created a conflict. Since `qux` doesn't contain a conflict, `jj diff` says that the conflict is resolved (which is what it always says if there is a conflict on the left side but not on the right side).

So, everything is "working as designed" AFAICT, and it's the design we'd have to adjust to help with this.

**Comment by martinvonz** [comment]
The `baz` diff is `bar=>baz` while the `qux` diff is `foo=>qux`. So the interdiff should somehow tell you that you're undoing the `bar=>baz` diff and doing the `foo=>qux` diff. Maybe #4062 would help. Instead of this:
```
Resolved conflict in foo:
   1     : <<<<<<< Conflict 1 of 1
   2     : %%%%%%% Changes from base to side #1
   3     : -bar
   4     : +foo
   5     : +++++++ Contents of side #2
   6     : baz
   7    1: >>>>>>> Conflict 1 of 1 endsqux
```

You would then get something like this:
```
Resolved conflict in foo:
  Part 1:
   1    1: fooqux # where "foo" is red and "qux" is green
  Part 2:
   1    1: bazbar # where "baz" is red and "bar" is green
```

Instead of just "Part 2" there, we could say something like "Discarded diff", but that might make it sound like the diff we display is the forward diff (i.e. `bar=>baz`), which it is not, and I don't think we should make it a forward diff either.

**Reply by ilyagr** [reply]
Here is what my prototype of #4062 😎 currently does.
<img width="519" alt="image" src="https://github.com/user-attachments/assets/b0ad7d53-9b59-4b32-ae86-7384aa4fa673">


Without colors:

```console
$ jj interdiff --from baz --to qux --conflicts
Resolved conflict in foo:
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< BEGIN CONFLICT DIFF
=========== Changed Add
-foo
+qux
=========== ---- Removed Diff
-| -bar
-| +baz
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>   END CONFLICT DIFF
```

@Zoybean , would this help at all?

In my mind, I'm not sure this kind of presentation is actually easy to understand to someone who is not really familiar with how conflicts work. But it should be useful for `blame`, `absorb`, and debugging complicated changes to conflicts when people want to do that (which should be rare).

If either of you have a suggestion for how to make it look clearer, let me know.  This is obviously not meant to be final (Some or all separators should be less loud, the ---- for "Removed Diff" will be gone, I just realized I forgot to remove it already. There are also important features missing, like line numbers).

(I am also using something closer to `diff3` presentation, mainly because I want it to be readable without colors, but borrow parts of color-words to make it more readable. Ultimately, we might have a presentation closer to color-words and a more machine-readable presentation, but currently it's a compromise that doesn't have all the advantages of either, but looks OK to me).

**Reply by ilyagr** [reply]
Perhaps the separators could say "======= Part 1: changed add", or some friendlier version of this. Though, I'm not sure there's a way to make it friendlier in a way that can also support "changed remove". Perhaps we can later have a "simple conflict diff mode" that skips confusing things like "changed remove", but that has its own problems.

**Reply by Zoybean** [reply]
I think that's clearer, yeah. It's at least helped me think about how to improve it.

The terminology of "Changed add" and "Removed diff" is still a bit confusing to me, and it took me rereading the sequence of operations that led to the interdiff, before I understood what the diff was saying. I was (incorrectly) interpreting "changed add" as the diff between the lines added by each change, and "removed diff" as the diff between the lines that were removed.

Spitballing ideas, how about:
- "New diff"/"Old diff"
- "Added diff"/"Removed diff"
- "Change added"/"Change removed"

```
Resolved conflict in foo:
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< BEGIN CONFLICT DIFF
=========== Change added
+| -foo
+| +qux
=========== Change removed
-| -bar
-| +baz
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>   END CONFLICT DIFF
```

Alternatively, what about prepending a marker to the ordinary diff lines, to indicate changes between diffs?
```
Resolved conflict in foo:
+| -foo
+| +qux
-| -bar
-| +baz
```
At a surface level, it seems to me like this could handle arbitrary levels of conflict-in-conflict somewhat cleanly, though I don't know how often that would come up, and I don't know if I'm missing any edge cases where it wouldn't work.

The "Resolved conflict in foo" still seems out of place in an interdiff to me, but the only improvements I can think of are unique to interdiffs and wouldn't apply to general conflicts.

**Reply by ilyagr** [reply]
> The "Resolved conflict in foo" still seems out of place in an interdiff to me, but the only improvements I can think of are unique to interdiffs and wouldn't apply to general conflicts.

👍 

> Spitballing ideas, how about:

The problem is that there are four different operations to describe. I think I'll rename them according to my description below:

- Adding a diff to a conflict, making it more complicated. For example, if you rebase one commit onto another, this would often add a diff to a conflict, turning an unconflicted commit into a conflicted one. (Or a conflicted commit into a conflict with more sides).
- Removing a diff from a conflict, making it less complicated. This is one of the things that happens in your example, and is the way conflicts get resolved. If we resolved the conflict to "foo", it would be the *only* thing that happens in the conflict diff.
- Changing a side of a conflict. This does not change the complexity (number of sides) of the conflict. This happens in our example.
- Changing a base of a conflict.  This does not change the complexity (number of sides) of the conflict either. I'm having a hard time coming up with a natural situation where this happens, but it seems necessary for completeness. Though, we could express such operations as a composition of removing a diff and adding a different diff, I suppose.

**Reply by ilyagr** [reply]
I'm not entirely happy with it, but here is what I have now:

<img width="510" alt="image" src="https://github.com/user-attachments/assets/c809ae99-31e1-49f6-9392-be0bdbfee15c">

**Reply by Zoybean** [reply]
> there are four different operations to describe.

Thanks, that breakdown clarified the complexity for me.

It looks like both changing a side and changing a base could be expressed as removing one diff and adding another. That would reduce conflicts and resolutions (and maybe all changes?) to combinations of 2 basic operations:
- adding a diff, (increasing the complexity, tending to cause conflict)
- removing a diff, (decreasing complexity, tending to resolve conflict)

Though, I'm not convinced that "simplifying" it down to 2 operations actually helps simplify the data presentation, so this probably isn't a useful realisation. Oh! Except that it seems to indicate that the hypothetical conflict-in-conflict I mentioned in an earlier comment, wouldn't really be a significant edge case at all, as it would still just be adding and removing ordinary diffs to/from a conflict, rather than some kind of complicated meta-conflict.

**Reply by ilyagr** [reply]
You can play with this if you want, #4251

**Reply by martinvonz** [reply]
> Oh! Except that it seems to indicate that the hypothetical conflict-in-conflict I mentioned in an earlier comment, wouldn't really be a significant edge case at all, as it would still just be adding and removing ordinary diffs to/from a conflict, rather than some kind of complicated meta-conflict.

Exactly! I actually stored higher-order conflicts as nested 2-sided conflicts in a very old version of jj (when it was still stored in a Mercurial repo, before I open-sourced it). I was happy when I realized it could be replaced by a list-based form because it made operations on conflicts a lot easier to reason about.

**Reply by martinvonz** [reply]
I think it might be best to try to keep the diffs in a conflict together as much as possible. So if we have a diff from `A+(B-C)` to `D+(E-F)`, I think it might be best to display that as diffs `D-A`, `E-F`, and `C-B` (negated `B-C` since it's on the left side). Maybe that's what you've already implemented, @ilyagr? I haven't checked your PR yet. If it's not, did you consider it and reject it?

**Reply by ilyagr** [reply]
> It looks like both changing a side and changing a base could be expressed as removing one diff and adding another. That would reduce conflicts and resolutions (and maybe all changes?) to combinations of 2 basic operations:
> That would reduce conflicts and resolutions (and maybe all changes?) to combinations of 2 basic operations:

I just realized that, up to conflict theory equivalence, "changed a side X -> Y" is equivalent to adding a diff with *base* X and *side* Y. If you then cancel out equivalent terms on top and bottom of the conflict, you get the same result as if you changed the side from X to Y.

I'm not sure whether this would lead to a clear way of presenting the situation to a user, but maybe? I'm not sure how to communicate the cancelling out of the terms.


---------------

For any `Z`, you could also present it as adding a diff Y-Z and a diff Z-X. For presentation, the empty `Z` could be OK to use, but generally the fact that we have to invent a `Z` that doesn't matter to split operations in this way doesn't seem that encouraging to me (even though it's a good point that this is possible).

**Reply by ilyagr** [reply]
> I think it might be best to try to keep the diffs in a conflict together as much as possible. So if we have a diff from `A+(B-C)` to `D+(E-F)`, I think it might be best to display that as diffs `D-A`, `E-F`, and `C-B` (negated `B-C` since it's on the left side). Maybe that's what you've already implemented, @ilyagr? I haven't checked your PR yet. If it's not, did you consider it and reject it?

The current implementation tries to optimize everything to minimize diff size (though the algorithms it currently uses is greedy and therefore not great).

I *think* that's generally better, especially for `blame` and `absorb` applications, but I have not re-thought this over in light of your comment and my previous comment, so remind me to mull it over again if I forget.

**Reply by ilyagr** [reply]
> I think it might be best to try to keep the diffs in a conflict together as much as possible. So if we have a diff from `A+(B-C)` to `D+(E-F)`, I think it might be best to display that as diffs `D-A`, `E-F`, and `C-B` (negated `B-C` since it's on the left side). Maybe that's what you've already implemented, @ilyagr? I haven't checked your PR yet. If it's not, did you consider it and reject it?

(Quoting myself ~[from Discord](https://discordapp.com/channels/968932220549103686/969291218347524238/1272075120919183404)~ **update:** decided to de-duplicate the post to just here)

Here are two equivalent presentations of the situation if the conflict 

```
<<<<<<< Conflict 1 of 1
%%%%%%% Changes from base to side #1
-bar
+foo
+++++++ Contents of side #2
baz
>>>>>>> Conflict 1 of 1 ends
```

was resolved to `foo`.

## Version 1

```
Resolved conflict in foo:
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< BEGIN CONFLICT DIFF
=========== Unchanged Side
foo
=========== Removed a Diff (-1 Base and -1 Side)
-| -bar
-| +baz
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>   END CONFLICT DIFF
```

and (equivalently)

## Version 2

```
Resolved conflict in foo:
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<< BEGIN CONFLICT DIFF
=========== Removed a Diff (-1 Base and -1 Side)
-| -bar
-| +foo
=========== Changed one of the Sides
-baz
+foo
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>   END CONFLICT DIFF
```

--------------

I'm not sure whether it would be good if we chose the latter presentation because it was closer to how `jj` materialized the original conflict. Or maybe? (Of course, we should also imagine a case where `foo` is much more complicated. Another case is  where `foo` is more complicated and was changed just a tiny bit)
